doc/sphinx/addendum/rewrite-rules.rst

```diff
@@ -0,0 +1,144 @@
+either by passing ``-allow-rewrite-rules`` to the
+Coq program or by turning the :flag:`Allow Rewrite Rules` flag on.
+
+.. flag:: Allow Rewrite Rules
+
+   This :term:`flag` enables the use of rewrite rules.
+   It is disabled by default and cannot be disabled once enabled.
+   The command-line flag ``-allow-rewrite-rules`` enables rewrite rules at
+   startup.
+
+   .. exn:: Rewrite rule declaration requires enabling the flag "Allow Rewrite Rules".
+      :undocumented:
+
+Declaring symbols
+-----------------
+
+Rewrite rules operate on symbols, which are their own kind of constants.
+
+.. cmd:: {| Symbol | Symbols } {| {+ ( @assumpt ) } | @assumpt }
+   :name: Symbol; Symbols
+
+   Symbols are declared in the same way as axioms, by giving their type and using the keyword ``Symbol``.
+
+   .. coqtop:: all
+
+      Set Allow Rewrite Rules.
+      Symbol pplus : nat -> nat -> nat.
+      Notation "a ++ b" := (pplus a b).
+
+Declaring rules
+---------------
+
+Rewrite rules have two parts named pattern (lhs) and production (rhs).
+When a rule is applied, the term is matched against the pattern, subterms aligned with pattern holes are collected
+and substitued into the production which is returned.
+
+.. cmd:: Rewrite {| Rule | Rules } @ident := @rewrite_rule {* with @rewrite_rule }
+  :name: Rewrite; Rule; Rules
+
+   Rewrite rules are declared in blocks, which bear a name.
+
+  .. prodn::
+     rewrite_rule ::= @term ==> @term
+
+
+  .. coqtop:: all
+
+     Rewrite Rule pplus_rewrite :=
+          ?n ++ 0 ==> ?n
+     with ?n ++ S ?m ==> S (?n ++ ?m)
+     with 0 ++ ?n ==> ?n
+     with S ?n ++ ?m ==> S (?n ++ ?m).
+
+Pattern syntax
+--------------
+
+Patterns are read from a subpart of terms, with the following syntax :
+
+  .. prodn::
+     pat ::= @qualid {? @univ_annot }
+     | @pat {+ @arg_pat }
+     | @pat .( @qualid {? @univ_annot } )
+     | match @pat {? as @name in @qualid {* {| @name | _ } } } {? return @arg_pat } with {? %| } {*| @qualid {* {| @name | _ } } => @arg_pat } end
+     | fun ({+ @name } {? : @arg_pat}) => @arg_pat
+     | forall ({+ @name } {? : @arg_pat}), @arg_pat
+     arg_pat ::= ? @name | _ | @pat
+
+where :n:`@qualid {? @univ_annot }` can refer to symbols, inductives and constructors, but not arbitrary constants.
+
+Equivalently, we can see a pattern as a pair of a rigid head and a list of eliminations :
+
+  .. prodn::
+     head_pat ::= @qualid {? @univ_annot }
+     | fun ({+ @name } {? : @arg_pat}) => @arg_pat
+     | forall ({+ @name } {? : @arg_pat}), @arg_pat
+     elimination ::= <> {+ @pat }
+     | <> .( @qualid {? @univ_annot } )
+     | match <> {? as @name in @qualid {* {| @name | _ } } } {? return @arg_pat } with {*| @qualid {* {| @name | _ } } => @arg_pat } end
+     pat ::= @head_pat {* @elimination }
+     arg_pat ::= ? @name | _ | @pat
+
+To be a valid pattern, it also must have a symbol as its head pattern, since the mechanism to trigger rewrite rules needs a symbol.
```

Comment:
```suggestion
Patterns must have a symbol (as defined by :cmd:`Symbol`) as its head pattern to be valid.
```

Is there an error message if there's no symbol?

Reply:
There is quite a lot of error messages that can be raised when the underlying term fails its parsing as a pattern, I don't think it would be interesting to document them